### PR TITLE
Pipeline persistance after crash

### PR DESCRIPTION
**Pull request by Shurc1val**
Added s3 storage of last INFORMATION log line to allow pipeline to continue if it crashed mid ride readings; also standardised 'user' to 'rider' across files used in the pipeline to solve an error that came up.

Will work on improving speed of the pipeline in a new branch.

### CODE REVIEW

File: pipeline/database_functions.py [hunk 1]
```diff
@@ -62,18 +62,18 @@ def select_address_from_database(db_connection : extensions.connection, address
...
         query = """INSERT INTO Rider(rider_id,address_id,first_name,last_name,
         birthdate,height,weight,email,gender,account_created)
           VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s) RETURNING rider_id;"""
 
-        parameters = (user['user_id'],user['address_id'],user['first_name'],
-                        user['last_name'],user['birthdate'],user['height'],
-                        user['weight'],user['email'],user['gender'],user['account_created'])
+        parameters = (rider['rider_id'],rider['address_id'],rider['first_name'],
+                        rider['last_name'],rider['birthdate'],rider['height'],
+                        rider['weight'],rider['email'],rider['gender'],rider['account_created'])
```
Comment (CharliedeanCS) [inline pipeline/database_functions.py:76]: Nicely done changing all the users to riders to keep consistently throughout the project

File: pipeline/pipeline.py [hunk 2]
```diff
@@ -5,20 +5,25 @@
...
 from dotenv import load_dotenv
 import load
 import transform
 import validate_heart_rate
-GROUP_ID = "testing23"
-EXTREME_HR_COUNT_THRESHOLD = 3
 
+GROUP_ID = "testing24"
+EXTREME_HR_COUNT_THRESHOLD = 3
+S3_BACKUP_FILENAME = "pipeline_backup.txt"
```
Comment (CharliedeanCS) [inline pipeline/pipeline.py:26]: Nice idea setting the filename as a constant. Makes it easy to change

File: pipeline/pipeline.py [hunk 3]
```diff
@@ -75,52 +80,96 @@ def ride_pipeline(log_line: str, bike_id: int) -> dict:
...
-    readings) upload to db and alert user by email if their heart rate has had an extreme value
+    readings) upload to db and alert rider by email if their heart rate has had an extreme value
     for enough consecutive readings.
     """
     reading = transform.get_reading_data_from_log_line(reading, log_line, start_time)
     if 'heart_rate' in reading:
         # Heart rate comes with the second of every pair of reading log lines.
-        if (reading['heart_rate'] == 0) or (user['min_heart_rate'] <= reading['heart_rate'] <= user['max_heart_rate']):
+        if (reading['heart_rate'] == 0) or \
+            (rider['min_heart_rate'] <= reading['heart_rate'] <= rider['max_heart_rate']):
```
Comment (CharliedeanCS) [inline pipeline/pipeline.py:94]: Is this whitespace in the actual code?

File: pipeline/pipeline.py [hunk 4]
```diff
@@ -75,52 +80,96 @@ def ride_pipeline(log_line: str, bike_id: int) -> dict:
...
+    a crash.
+    """
+    try:
+        s3_client = client("s3",
+                        aws_access_key_id=environ['AWS_ACCESS_KEY_ID_'],
+                        aws_secret_access_key=environ['AWS_SECRET_ACCESS_KEY_'])
+
+        s3_client.put_object(Body = log_line, Bucket = environ['BUCKET_NAME'],
+                            Key = filename)
+    except ClientError:
```
Comment (CharliedeanCS) [inline pipeline/pipeline.py:122]: Nice use of try and Catch to ensure the pipeline stays running.

File: pipeline/pipeline.py [hunk 5]
```diff
@@ -75,52 +80,96 @@ def ride_pipeline(log_line: str, bike_id: int) -> dict:
...
+        pass
+
+
+def retrieve_text_from_s3_file(filename: str = S3_BACKUP_FILENAME):
+    """Retrieves body of text file stores in s3_bucket."""
+    try:
+        s3_client = client("s3",
+                           aws_access_key_id=environ['AWS_ACCESS_KEY_ID_'],
+                           aws_secret_access_key=environ['AWS_SECRET_ACCESS_KEY_'])
```
Comment (CharliedeanCS) [inline pipeline/pipeline.py:131]: I see you have repeated lines with creating a s3_client. Maybe creating the s3 client could be a function?

File: pipeline/pipeline.py [hunk 6]
```diff
@@ -75,52 +80,96 @@ def ride_pipeline(log_line: str, bike_id: int) -> dict:
...
+                ride = ride_pipeline(system_log_line, bike_id)
+                reading = {'ride_id': ride['ride_id']}
+
+            first_relevant_line = False
+
+        if ('[INFO]' in log_line) and rider:
             reading = reading_pipeline(
-                log_line, ride['ride_id'], ride['start_time'], reading, user, consecutive_extreme_hrs)
+                log_line, ride['ride_id'], ride['start_time'], reading, rider,
+                consecutive_extreme_hrs)
```
Comment (CharliedeanCS) [inline pipeline/pipeline.py:172]: Nice fix to ensure the pipeline keeps running even when it starts up halfway through a ride.

File: pipeline/pipeline.py [hunk 7]
```diff
@@ -75,52 +80,96 @@ def ride_pipeline(log_line: str, bike_id: int) -> dict:
...
-        elif ('[INFO]' in log_line) and (not new_ride):
+
+        if ('[SYSTEM]' in log_line) or (('[INFO]: Ride' in log_line) and first_relevant_line):
+            # SYSTEM log_line, or the pipeline is starting mid ride.
+
+            if '[INFO]' in log_line:
+                system_log_line = retrieve_text_from_s3_file()
+            else:
+                system_log_line = log_line
+                save_log_line_to_s3(system_log_line)
```
Comment (CharliedeanCS) [inline pipeline/pipeline.py:156]: Very nice solution

File: pipeline/test_pipeline.py [hunk 8]
```diff
@@ -225,8 +238,262 @@ def test_pipeline(mock_load_bike, mock_transform_bike, mock_reading_pipeline, mo
...
+
+@patch('pipeline.save_log_line_to_s3')
+@patch('pipeline.retrieve_text_from_s3_file')
+@patch('pipeline.get_next_log_line')
+@patch('pipeline.get_kafka_consumer')
+@patch('pipeline.rider_pipeline')
+@patch('pipeline.ride_pipeline')
+@patch('pipeline.reading_pipeline')
+@patch('transform.get_bike_serial_number_from_log_line')
+@patch('load.add_bike')
```
Comment (CharliedeanCS) [inline pipeline/test_pipeline.py:320]: Nice patch job

File: pipeline/pipeline.py (hunk 6 above)
Comment (lcawood) [reply pipeline/pipeline.py:187]: I agree, nice work!

**Review by lcawood — APPROVED** [review]
I fixed the conflict quickly as I updated the gender issue afterwards and this all looks good to me! :)

**Review by lcawood — APPROVED** [review]
Hard to see which changes are the ones not 'user' -> 'rider' haha, tests are very comprehensive and really impressed with the S3 solution as a whole, well done on this!!

**Pull request merged**